src/python/turicreate/toolkits/object_detector/_tf_image_augmenter.py

```diff
@@ -22,38 +22,267 @@
-    return isinstance(x, (ops.Tensor, variables.Variable))
+        if max_saturation is not None and max_saturation > 1.0:
+            log_sat = np.log(max_saturation)
+            image = tf.image.random_saturation(image, lower=np.exp(-log_sat), upper=np.exp(log_sat))
 
-def image_dimensions(images, static_only=True):
-    # Returns the dimensions of an image tensor.
-    if static_only or images.get_shape().is_fully_defined():
-        return images.get_shape().as_list()
-    else:
-        return tf.unstack(tf.shape(images))
+        if max_contrast is not None and max_contrast > 1.0:
+            log_con = np.log(max_contrast)
+            image = tf.image.random_contrast(image, lower=np.exp(-log_con), upper=np.exp(log_con))
+        
+        image = tf.clip_by_value(image, 0, 1)
+        colored_images.append(image)
+    return colored_images, annotations
+
+def _is_intersection(bounding_box1, bounding_box2):
+    bounding_box1_unstacked = tf.unstack(bounding_box1)
+    bounding_box2_unstacked = tf.unstack(bounding_box2)
+    is_intersection_cond = math_ops.logical_and(math_ops.logical_and(math_ops.less(bounding_box1_unstacked[0], bounding_box2_unstacked[2]),
+                                                                     math_ops.less(bounding_box1_unstacked[1], bounding_box2_unstacked[3])),
+                                                math_ops.logical_and(math_ops.less(bounding_box2_unstacked[0], bounding_box1_unstacked[2]),
+                                                                     math_ops.less(bounding_box2_unstacked[1], bounding_box1_unstacked[3])))
+    
+    return control_flow_ops.cond(is_intersection_cond,
+                                lambda: True,
+                                lambda: False)
+def _get_middle_points(bounding_box1, bounding_box2):
+    stacked_points = tf.transpose(tf.stack([bounding_box1, bounding_box2]))
+    uts = tf.unstack(stacked_points)
+
+    flip_values_0 = math_ops.greater(uts[0][1], uts[0][0])
+    uts[0] = control_flow_ops.cond(flip_values_0,
+                        lambda: array_ops.reverse(uts[0], [0]),
+                        lambda: uts[0])
+    
+    flip_values_1 = math_ops.greater(uts[1][1], uts[1][0])
+    uts[1] = control_flow_ops.cond(flip_values_1,
+                        lambda: array_ops.reverse(uts[1], [0]),
+                        lambda: uts[1])
+    
+    flip_values_2 = math_ops.less(uts[2][1], uts[2][0])
+    uts[2] = control_flow_ops.cond(flip_values_2,
+                        lambda: array_ops.reverse(uts[2], [0]),
+                        lambda: uts[2])
+    
+    flip_values_3 = math_ops.less(uts[3][1], uts[3][0])
+    uts[3] = control_flow_ops.cond(flip_values_3,
+                        lambda: array_ops.reverse(uts[3], [0]),
+                        lambda: uts[3])
+
+    restack_points = tf.stack(uts)[:, :1]
+    reshaped_points = tf.reshape(restack_points, [ -1, 4])
+    
+    return tf.squeeze(reshaped_points)
```

Comment:
If there is only a single bounding box, couldn't this oversqueezed? I always prefer to have an explicit axis defined when squeezing to prevent corner cases.

Reply:
I guess this function only operates on a single box?

I feel like this operation can be done vectorized for batches, though.